controlplane/kubeadm/internal/filters.go

```diff
@@ -391,13 +391,4 @@ func cleanupConfigFields(kcpConfig *bootstrapv1.KubeadmConfigSpec, machineConfig
 		machineConfig.Spec.JoinConfiguration != nil {
 		machineConfig.Spec.JoinConfiguration.NodeRegistration = emptyNodeRegistration
 	}
-
-	// Clear up the TypeMeta information from the comparison.
```

Comment:
Can we give it a quick try if we can revert
```go
	// Init configuration is usually be set under normal circumstances, but with some test providers (e.g. CAPD in memory)
	// it is left empty; In this case, we are initializing it so the comparison code can be simplified.
	if kcpConfig.InitConfiguration == nil {
		kcpConfig.InitConfiguration = &bootstrapv1.InitConfiguration{}
	}
```
in getAdjustedKcpConfig now?

(you'll have to rebase first, it doesn't exist on this PR yet)